src/libcore/slice/mod.rs

```diff
@@ -1575,6 +1575,127 @@ impl<T> [T] {
         sort::quicksort(self, |a, b| f(a).lt(&f(b)));
     }
 
+    /// Rearranges the elements of the slice such that the following properties hold:
+    ///
+    /// - The value at position `index` will be the same as if the array were sorted.
+    /// - Any value at position `i < index` will be less than or equal to any value at a position
+    ///   `j > index`.
+    ///
+    /// This reordering is unstable (i.e. any number of equal elements may end up at position
+    /// `index`), in-place (i.e. does not allocate), and `O(n)` worst-case.
+    ///
+    /// # Current implementation
+    ///
+    /// The current algorithm is based on the quickselect portion of the same quicksort algorithm
+    /// used for [`sort_unstable`].
+    ///
+    /// [`sort_unstable`]: #method.sort_unstable
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(sort_at_index)]
+    ///
+    /// let mut v = [-5i32, 4, 1, -3, 2];
+    ///
+    /// v.sort_at_index(2);
+    /// assert!(v == [-3, -5, 1, 2, 4] ||
+    ///         v == [-5, -3, 1, 2, 4] ||
+    ///         v == [-3, -5, 1, 4, 2] ||
+    ///         v == [-5, -3, 1, 4, 2]);
+    /// ```
+    #[unstable(feature = "sort_at_index", issue = "55300")]
+    #[inline]
+    pub fn sort_at_index(&mut self, index: usize)
```

Comment:
Why not have this API return the element at the index, save the user a boundary check?

Reply:
Could you elaborate how this would look? The caller will already have to have borrowed the vector mutably to call the function in the first place.